presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java

```diff
@@ -321,13 +322,31 @@ public boolean equals(Object o)
 
         TypeSignature other = (TypeSignature) o;
 
+        // TODO remove this hack together with hack from toString()
+        if (magicVarcharEquals(other, this)) {
+            return true;
+        }
+
         return Objects.equals(this.base.toLowerCase(Locale.ENGLISH), other.base.toLowerCase(Locale.ENGLISH)) &&
                 Objects.equals(this.parameters, other.parameters);
     }
 
+    private boolean magicVarcharEquals(TypeSignature b, TypeSignature a)
```

Comment:
Changed to `first` and `second`. Also fixed the bug, as `this` was used instead of `first` in two places.